drivers: display: ssd1322 driver

topisani opened the pull request:
This is based on https://github.com/zephyrproject-rtos/zephyr/pull/68608, but adds more configuration options

Most options map directly to values documented in the datasheet, except segments-per-pixel, which I had to add to support the Newhaven NHD-2.7-12864WDW3. This is a slightly odd feature, but in practice it is a lot nicer to support it in the driver, and since we're currently remapping pixels anyway, it makes sense to implement here in my opinion.

This commit also adds a configurable buffer size for the pixel conversion. By using a larger buffer, we can potentially use DMA for the SPI transfer. The default is set to 8, which is the smallest value that supports `segments-per-pixel = 2`

decsny commented on dts/bindings/display/solomon,ssd1322.yaml:
```diff
@@ -0,0 +1,63 @@
+
+  row-offset:
+    type: int
+    default: 0
```
https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#rules-for-default-values

danieldegrasse commented on drivers/display/ssd1322.c:
```diff
@@ -28,10 +29,20 @@ struct ssd1322_config {
 
-static inline int ssd1322_write_command(const struct device *dev, uint8_t cmd,
-					const uint8_t *buf, size_t len)
+static inline int ssd1322_write_command(const struct device *dev, uint8_t cmd, const uint8_t *buf,
```
Please do not make formatting changes like this in the same commit as functional additions. If you would like to format a file before editing, IE with clang-format, then please do so and keep the formatting changes in *one commit*. This makes the review process much easier

danieldegrasse reviewed (commented):
Generally looks ok, (although the default values do need justification).

@hlukasz would you prefer to merge your PR first once the merge window opens, then have this PR contain only the enhancements? Or would it be better for you to merge both commits within this PR? By default I'd bias towards merging #68608 first, but it would be more efficient to batch the commits into one PR

hlukasz commented:
I don't have strong opinion on that. If we can merge both patches as one MR in next merging window, let's go that way.

danieldegrasse commented:
> I don't have strong opinion on that. If we can merge both patches as one MR in next merging window, let's go that way.

Ok, we can merge both in one PR then. @topisani when you get a chance- it probably makes the most sense to squash the driver itself into one commit, since the driver introduction and extensions are happening in one PR. Please make sure that @hlukasz is a co-author on the commit, and that copyrights are retained- I want to be sure credit is present where it is due :) This also allows us to avoid the clang-format changes, since the driver changes will all be in one commit

danieldegrasse commented:
@jfischer-no would you be willing to review this PR? I know you had some comments on #68608, and I'm wondering if we could them resolved here to get this driver moving forwards

jfischer-no commented:
> @jfischer-no would you be willing to review this PR? I know you had some comments on #68608, and I'm wondering if we could them resolved here to get this driver moving forwards

Why should we accept PRs that try to push something like `cmd_data[1] = (y + desc->height - 1);` or variables that are dropped in the middle of a block for no reason, or a header file with effectively only 20 lines that is not and will never be used anywhere else, when the author has been told to fix/improve it?

topisani commented:
@jfischer-no I have attempted to address your review on the original PR.

jfischer-no commented on drivers/display/ssd1322.c:
```diff
@@ -0,0 +1,353 @@
+	}
+
+	/*
+	 * Controller uses 4-bit grayscale format, so one pixel is represented by 4 bits.
+	 * Zephyr's display API does not support this format, so I am using 1-bit b/w mode and
+	 * convert each 1-bit pixel to 1111 or 0000.
```
Maybe as followup idea, perhaps it makes more sense, or is easier, to map e.g. RGB_888 to 4-bit grayscale.

topisani replied:
Agreed. The restructured conversion logic in my latest revision should make it fairly easy to add support for other formats in the future.

jfischer-no commented on drivers/display/Kconfig.ssd1322:
```diff
@@ -0,0 +1,22 @@
+
+if SSD1322
+
+config SSD1322_BUFFER_SIZE
+	int "SSD1322 internal buffer size"
+	default 8
+	range 8 61440
+	help
+	  Size of the SSD1322 internal pixel format conversion buffer, in bytes.
+	  Maximum value is the size of the display ram, 480*128=61440.
```
There is already information in devicetree what buffer size is needed. So you could add
```c
struct ssd1322_data {
    uint8_t *const buf;
}
```
and define a buffer based on `DT_PROP(node_id, height)....` and assign it to the pointer. This will also give you better multi-instance support. Also maximum size is to much, good compromise should be buffer large enough for line.

jfischer-no commented on drivers/display/ssd1322.c:
```diff
@@ -0,0 +1,353 @@
+		uint16_t seg_idx = 0;
+		uint8_t color;
+
+		while (in_idx < pixel_count && seg_idx / 2 < sizeof(conversion_buf)) {
+			in_idx = seg_idx / config->segments_per_pixel;
+			color = (buf_iter[in_idx / 8] & BIT(in_idx % 8)) ? 0xF : 0;
+			conversion_buf[seg_idx / 2] =
+				(seg_idx % 2 == 0) ? color
+						   : ((color << 4) | conversion_buf[seg_idx / 2]);
+			seg_idx++;
+		}
```
It does not look nice, maybe you can move it to a separate e.g. mono01_to_grayscale conversion function.

topisani replied:
Agreed. I restructured the whole conversion logic to be slightly more readable.

jfischer-no reviewed (commented):
LGTM, just few comments

topisani commented on drivers/display/ssd1322.c:
```diff
@@ -0,0 +1,380 @@
+ *
+ * @param in_buf: pointer to input buffer in mono01 format. This value will bel updated to point to
+ * the first byte after the last converted pixel.
+ * @param pixel_count: pointer to the total number of pixels in in_buf to convert. The number of
```
The compliance check fails with a typo error here. How do i address that?

jfischer-no replied:
Name it buf_in and buf_out :-). Btw, please do not use doxygen style here.

topisani replied:
nice workaround ;-)

topisani committed:
drivers: display: Add ssd1322 driver

Initial support for SSD1322 OLED display driver. Only 1 bit color mode is
supported.

Most options map directly to values documented in the datasheet,
except segments-per-pixel, which I had to add to support the Newhaven
NHD-2.7-12864WDW3. This is a slightly odd feature, but in practice
it is a lot nicer to support it in the driver, and since we're currently
remapping pixels anyway, it makes sense to implement here.

This driver also has a configurable buffer size for the pixel conversion.
By using a larger buffer, we can potentially use DMA for the SPI transfer.
The default is set to 8, which is the smallest value that supports
segments-per-pixel = 2

Initial driver implementation by Lukasz Hawrylko <lukasz@hawrylko.pl>.
Additional options and configurability by Tobias Pisani <mail@topisani.dev>

Signed-off-by: Lukasz Hawrylko <lukasz@hawrylko.pl>
Signed-off-by: Tobias Pisani <mail@topisani.dev>

Co-authored-by: Lukasz Hawrylko <lukasz@hawrylko.pl>
Co-authored-by: Tobias Pisani <mail@topisani.dev>

jfischer-no commented:
@topisani There are conflicts you need to rebase on latest main.

topisani committed:
tests: drivers: display: build_all: add ssd1322

Add ssd1322 to display build test

Signed-off-by: Tobias Pisani <mail@topisani.dev>
